### PR TITLE
refactor s3 tests

### PR DESCRIPTION
**Pull request by bentsku**
This is the initial PR for refactoring s3 tests, adding new ones corresponding to issues we could/would fix while migrating the service to ASF.

- [x] Migrated tests from `integration/test_s3.py` to `integration/s3/test_s3.py` 
   - only a few left, where we still need to decide if migration makes sense - will takle during ASF s3 migration
- [x] Validated against AWS and added markers; added snapshots for appropriate tests 
- [x] Added a few tests corresponding to opened issue, marked with xfail

### CODE REVIEW

File: localstack/testing/pytest/fixtures.py [hunk 1]
```diff
@@ -198,6 +199,41 @@ def s3_client() -> "S3Client":
...
+        )
+        if os.environ.get("TEST_DISABLE_RETRIES_AND_TIMEOUTS")
+        else None
+    )
+    return aws_stack.connect_to_service(
+        "s3",
+        config=boto_config,
+        aws_access_key_id=TEST_AWS_ACCESS_KEY_ID,
+        aws_secret_access_key=TEST_AWS_SECRET_ACCESS_KEY,
+    )
```
Comment (dominikschubert) [inline localstack/testing/pytest/fixtures.py:234]: small heads-up: 
these can probably be simplified with #6824

Comment (bentsku) [reply localstack/testing/pytest/fixtures.py:234]: Awesome! We also have some of those inside `integration/s3/test_s3.py`, that is going to be super useful! Thanks! 🙏

File: tests/integration/s3/test_s3.py [hunk 2]
```diff
@@ -245,175 +516,2240 @@ def test_put_object_checksum(self, s3_client, s3_create_bucket, algorithm):
...
+        object_key_expired = "key-object-expired"
+        object_key_not_expired = "key-object-not-expired"
+
+        s3_client.put_object(
+            Bucket=s3_bucket,
+            Key=object_key_expired,
+            Body="foo",
+            Expires=short_expire,
         )
+        time.sleep(3)
```
Comment (alexrashed) [inline tests/integration/s3/test_s3.py:682]: A comment for the sleep would be nice (like below `# sleep so it expires`).

File: tests/integration/s3/test_s3.py [hunk 3]
```diff
@@ -245,175 +516,2240 @@ def test_put_object_checksum(self, s3_client, s3_create_bucket, algorithm):
...
+            retries=10,
+            sleep=1,
+            function_name=function_name,
+            regex_filter="success",
+            expected_length=1,
+            logs_client=logs_client,
+        )
+
+    @pytest.mark.aws_validated
+    # TODO LocalStack adds this RequestID to the error response
```
Comment (alexrashed) [inline tests/integration/s3/test_s3.py:1170]: Does AWS not have a request ID?

File: tests/integration/s3/test_s3.py [hunk 4]
```diff
@@ -245,175 +516,2240 @@ def test_put_object_checksum(self, s3_client, s3_create_bucket, algorithm):
...
+            Bucket=bucket, Key=object_key, Body="<h1>Index</html>", ACL="public-read"
+        )
+
+        response = requests.get(
+            url, headers={"Origin": config.get_edge_url(), "Content-Type": "text/html"}
+        )
+        assert 200 == response.status_code
+        # TODO this is not contained in AWS but was asserted for LocalStack
+        #  assert "Access-Control-Allow-Headers" in response.headers
+        #  assert "x-amz-tagging" == response.headers["Access-Control-Allow-Headers"]
```
Comment (alexrashed) [inline tests/integration/s3/test_s3.py:2484]: If these are not contained in AWS, I think we should just remove this here.

File: tests/integration/s3/test_s3.py [hunk 5]
```diff
@@ -245,175 +516,2240 @@ def test_put_object_checksum(self, s3_client, s3_create_bucket, algorithm):
...
+
+        # TODO this does not work on AWS (not even if the first delete was not executed)
+        # presigned_delete_url_2 = _generate_presigned_url(
+        #     client,
+        #     {"Bucket": bucket_name, "Key": object_key, "VersionId": "1"},
+        #     expires,
+        #     client_method="delete_object",
+        # )
+        # response = requests.delete(presigned_delete_url_2)
+        # assert 204 == response.status_code
```
Comment (alexrashed) [inline tests/integration/s3/test_s3.py:2740]: Same as above, I think this section can just be removed.

**Review by alexrashed — APPROVED** [review]
Wow! This is huge! It's really great to see such a big AWS-validated / snapshot test coverage for one of the absolute core services! 🥇 🔝 
I only found some minor things in here, where I'm not completely sure if these things weren't just migrated.

File: tests/integration/s3/test_s3.py (hunk 3 above)
Comment (bentsku) [reply tests/integration/s3/test_s3.py:1170]: No, AWS does not add it, we realised this with the snapshots I believe, we encountered this issue multiple times.

File: tests/integration/s3/test_s3.py [hunk 6]
```diff
@@ -245,175 +516,2240 @@ def test_put_object_checksum(self, s3_client, s3_create_bucket, algorithm):
...
+        response["DeleteResult"].pop("@xmlns")
+        # assert response["DeleteResult"]["Error"]["Key"] == object_key_1
+        # assert response["DeleteResult"]["Error"]["Code"] == "AccessDenied"
+        # assert response["DeleteResult"]["Deleted"]["Key"] == object_key_2
+        snapshot.match("multi-delete-with-requests", response)
+
+        response = s3_client.list_objects(Bucket=bucket_name)
+        snapshot.match("list-remaining-objects", response)
+        # assert 200 == response["ResponseMetadata"]["HTTPStatusCode"]
+        # assert len(response["Contents"]) == 1
```
[the remark this answers is not in the document]
Comment (bentsku) [reply tests/integration/s3/test_s3.py:1926]: Oops, will do, thanks!

File: tests/integration/s3/test_s3.py [hunk 7]
```diff
@@ -156,40 +293,28 @@ def test_list_objects_with_prefix(self, s3_client, s3_create_bucket, delimiter):
...
-        response = e.value.response
-        assert response["ResponseMetadata"]["HTTPStatusCode"] == 404
-        error = response["Error"]
-        assert error["Code"] == "NoSuchBucket"
-        assert error["Message"] == "The specified bucket does not exist"
+        snapshot.match("expected_error", e.value.response)
 
     @pytest.mark.aws_validated
-    def test_delete_bucket_no_such_bucket(self, s3_client):
+    @pytest.mark.skip_snapshot_verify(path="$..RequestID")
```
[the remark this answers is not in the document]
Comment (bentsku) [reply tests/integration/s3/test_s3.py:304]: It's divergence, LocalStack includes the `RequestID` in the error response but AWS does not. It must come from moto

File: tests/integration/s3/test_s3.py [hunk 8]
```diff
@@ -156,40 +293,28 @@ def test_list_objects_with_prefix(self, s3_client, s3_create_bucket, delimiter):
         assert is_sub_dict(sub_dict, response)
 
     @pytest.mark.aws_validated
-    def test_get_object_no_such_bucket(self, s3_client):
+    @pytest.mark.skip_snapshot_verify(path="$..Error.BucketName")
```
[the remark this answers is not in the document]
Comment (steffyP) [reply tests/integration/s3/test_s3.py:296]: It _should_. But somehow, the current s3_listener uses different approaches to throw `NoSuchBucket` errors. For the `GetBucket` case it doesn't set the `BucketName` in the first place. We need to revise all `skip_snapshot_verify` paths, for the new s3-asf provider :)

File: tests/integration/s3/test_s3.py (hunk 3 above)
Comment (alexrashed) [reply tests/integration/s3/test_s3.py:1159]: As mentioned above, let's take a look once the new provider is in place. :)

**Pull request merged**